scripts/tests/forest_cli_check.sh

```diff
@@ -1,45 +1,70 @@
-
-echo "Fetching params"
-$FOREST_CLI_PATH fetch-params --keys
-echo "Downloading zstd compressed snapshot without aria2"
-$FOREST_CLI_PATH --chain calibnet snapshot fetch --provider filecoin --compressed -s "$SNAPSHOT_DIRECTORY"
-echo "Downloading snapshot without aria2"
-$FOREST_CLI_PATH --chain calibnet snapshot fetch -s "$SNAPSHOT_DIRECTORY"
-echo "Cleaning up snapshots"
-$FOREST_CLI_PATH --chain calibnet snapshot clean -s "$SNAPSHOT_DIRECTORY" --force
-echo "Cleaning up snapshots again"
-$FOREST_CLI_PATH --chain calibnet snapshot clean -s "$SNAPSHOT_DIRECTORY" --force
-echo "Downloading zstd compressed snapshot"
-$FOREST_CLI_PATH --chain calibnet snapshot fetch --aria2 --provider filecoin --compressed -s "$SNAPSHOT_DIRECTORY"
-echo "Cleaning up database"
-$FOREST_CLI_PATH --chain calibnet db clean --force
-echo "Cleaning up database again"
-$FOREST_CLI_PATH --chain calibnet db clean --force
-
-echo "Downloading snapshot"
-$FOREST_CLI_PATH --chain calibnet snapshot fetch --aria2 -s "$SNAPSHOT_DIRECTORY"
-
-echo "Validating as mainnet snapshot"
-set +e
-$FOREST_CLI_PATH --chain mainnet snapshot validate "$SNAPSHOT_DIRECTORY"/*.car --force && \
-{
-    echo "mainnet snapshot validation with calibnet snapshot should fail";
-    exit 1;
+snapshot_dir=$(mktemp --directory)
+
+# return the first snapshot path listed by forest
+function pop-snapshot-path() {
+    forest-cli snapshot list --snapshot-dir "$snapshot_dir" \
+        | grep path \
+        | sed 's/^\tpath: //' \
+        | head --lines=1
+}
+
+function clean-snapshot-dir() {
+    rm --recursive --force --verbose -- "$snapshot_dir"
+    mkdir --parents -- "$snapshot_dir"
 }
-set -e
 
-echo "Validating as calibnet snapshot (uncompressed)"
-$FOREST_CLI_PATH --chain calibnet snapshot validate "$SNAPSHOT_DIRECTORY"/*.car --force
+# TODO(aatifsyed): I don't really understand what this does
+"$FOREST_CLI_PATH" fetch-params --keys
+
+: fetch snapshot
+"$FOREST_CLI_PATH" --chain calibnet snapshot fetch --snapshot-dir "$snapshot_dir"
+clean-snapshot-dir
+
+
+: "clean database (twice)"
+# TODO(aatifsyed): I don't really understand what this tests
+"$FOREST_CLI_PATH" --chain calibnet db clean --force
+"$FOREST_CLI_PATH" --chain calibnet db clean --force
+
+: validate calibnet snapshot
+    clean-snapshot-dir
+
+    : : fetch a calibnet snapshot
+    "$FOREST_CLI_PATH" --chain calibnet snapshot fetch --snapshot-dir "$snapshot_dir"
+    validate_me=$(pop-snapshot-path)
+
+    : : validating under calibnet chain should succeed
+    "$FOREST_CLI_PATH" --chain calibnet snapshot validate "$validate_me" --force
+
+    : : validating under mainnet chain should fail
+    if "$FOREST_CLI_PATH" --chain mainnet snapshot validate "$validate_me" --force; then
+        exit 1
+    fi
+
+    clean-snapshot-dir
+
+: intern a snapshot downloaded with aria2c
+    download_dir=$(mktemp --directory)
+    aria2c https://forest.chainsafe.io/calibnet/snapshot-latest.car.zst --dir="$download_dir"
```

Comment:
Is the Forest fetch no longer using `aria2c` if possible?

Reply:
Correct - if users want to use aria2c, they could (should?) do that and then `intern`.

Note a quick benchmark found no statistically significant different between the two, at least for the calibnet snapshot (n=20)

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `reqwest` | 9.045 ± 3.620 | 6.818 | 20.194 | 1.04 ± 0.51 |
| `aria2c` | 8.693 ± 2.491 | 5.722 | 14.846 | 1.00 |

```bash
hyperfine \
    --command-name=reqwest \
    './target/release/benchmark --out=/dev/null https://forest.chainsafe.io/calibnet/snapshot-latest.car.zst' \
    --command-name=aria2c \
    'aria2c --dir=/dev --out=null --file-allocation=none --allow-overwrite=true https://forest.chainsafe.io/calibnet/snapshot-latest.car.zst' \
    --export-markdown=benchmark.md \
    --export-json=benchmark.json \
    --export-csv=benchmark.csv \
    --show-output \
    --runs=20
```

```rust
use std::path::PathBuf;

use clap::Parser as _;
use futures::TryStreamExt as _;
use indicatif::ProgressBar;
use tap::Pipe as _;
use tokio::fs::File;
use url::Url;

#[derive(clap::Parser)]
pub struct Args {
    pub url: Url,
    #[arg(short, long)]
    pub out: PathBuf,
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let progress_bar = ProgressBar::new(0);
    let mut dst = File::create(args.out).await?;
    let mut src = reqwest::get(args.url)
        .await?
        .error_for_status()?
        .bytes_stream()
        .map_err(|reqwest_error| std::io::Error::new(std::io::ErrorKind::Other, reqwest_error))
        .pipe(tokio_util::io::StreamReader::new)
        .pipe(|it| progress_bar.wrap_async_read(it));
    tokio::io::copy(&mut src, &mut dst).await?;
    Ok(())
}
```

Reply:
You missing the `--split=5` argument to aria2c. With this flag, `aria2c` is more than twice as fast with my internet (1Gbit/s wired connection). The difference is even bigger when using wifi. And in India and China, the difference should be bigger still.

Reply:
If we follow this path, we could argue they could do it with `wget` on their machine anyway. As @lemmih noted, `aria2c` is significantly faster with correct settings, especially when downloading massive mainnet snapshots. I do believe it also supports resuming failed/interrupted downloads, which we supported with the `aria2c` path as far as I remember (but not native).

If anything, I'd consider having `aria2c` as a necessary dependency for Forest. That would annoy potential users that would get a runtime error, though.

Reply:
ack, adding back